### DISCUSSION TITLE
Get Total book page count, read time per page & number of pages read on android

### DISCUSSION THREAD

**Discussion by ANasser-VLink** [readium/kotlin-toolkit#52] [Q&A]
Greetings I would like to inquire about the use of Readium as an EPUB reader with LCM. 
First of all I'm kind of lost to which Repo i should be using, I've been using "int/testapp-with-lcp" so please correct me If I'm using the wrong version.

I am working on a demo to see if we can integrate our app with Readium and reach this state with these features:
 - The app needs access to the current page the user is reading 
 - Calculate how much time is spent in each page. 
 - Besides getting the total number of pages in a book when it's opened. 
 - Access the index of the page when the users re-opens a book to resumes reading the book.
I can't find a way to interact with the navigator and the streamer as they are used as libraries without access to source code. so as a workaround i downloaded and imported their code locally.
Am I missing something as in an interface for the onTap, onSwipe (left and right) and onPageChanged that I can implement to handle these actions?

Sorry that the post is long and a bit unorganized.

**Comment by mickael-menu** [comment]
> EPUB reader with LCM.

I assume you meant LCP?

> First of all I'm kind of lost to which Repo i should be using, I've been using "int/testapp-with-lcp" so please correct me If I'm using the wrong version.

Yes, that's the correct one as an integrator, if you don't want to modify the sources of Readium. If you want to use Readium with Git submodules instead, to have access to the sources, you can try `dev/testapp-with-lcp`. 

> * The app needs access to the current page the user is reading

We don't have the concept of "screen pages" in Readium (or in EPUB for that matter), so this info is not available. However we have `position` in our `Locator` model which can be used as an alternative, see: https://github.com/readium/architecture/tree/master/models/locators

The pros of positions compared to page numbers is that they don't change with screen ratio or user settings, and so can be used to share the same positions across devices. See `Navigator.currentLocator` in `r2-navigator-kotlin`.

> * Calculate how much time is spent in each page.

This will be something to handle by your app. You could monitor changes of `Navigator.currentLocator`.

> * Besides getting the total number of pages in a book when it's opened.

`publication.positions.count`, but they are not "screen page numbers"

> * Access the index of the page when the users re-opens a book to resumes reading the book.

This is handled by saving the `Navigator.currentLocator` and restoring it as the `initialLocator` parameter of the navigators. See the example in `r2-testapp-kotlin`.

>   Am I missing something as in an interface for the onTap, onSwipe (left and right) and onPageChanged that I can implement to handle these actions?

`onSwipe` to turn pages is handled by the navigator itself. You can override `onTap` though, see the `Navigator` interface in `r2-navigator-kotlin`.

**Comment by paulocoutinhox** [comment]
Hi @mickael-menu,

I have the same questions and i still don't understand how to get some infos, or maybe it is not available.

- Current resource HTML: `val newHREF = navigator.currentLocator.value.href`
- Current  position (can be use to show current page number): `val newPageIndex = navigator.currentLocator.value.locations.position ?: 0`

But:

- Current total positions (global): `??`
- Current total positions (for current resource): `??`

Can you help me?

**Reply by paulocoutinhox** [reply]
Total global positions for publication can be `publication.positions().count()`?

So, current position is globally too?

Example:

```kotlin
val globalTotal = publication.positions().count()
val globalPageIndex = navigator.currentLocator.value.locations.position
```

Is this? If yes, i don't need know in what resource it is, correct?

**Reply by mickael-menu** [reply]
```kotlin
// Number of positions in the whole publication.
val positionCount = publication.positions().count()

// Current position.
val currentPosition = navigator.currentLocator.value.locations.position

val currentResourceHref = navigator.currentLocator.value.href
// Index of the current resource in the reading order.
val currentResourceIndex = publication.readingOrder.indexOfFirstWithHref(currentResourceHref)
// List of positions in the resource.
val positionsInResource = publication.positionsByReadingOrder()[currentResourceIndex]
// Number of positions in the resource.
val positionCountInResource = positionsInResource.count()
// Number of the first position in the resource.
val firstPositionInResource = positionsInResource.firstOrNull()?.locations?.position
// Number of the last position in the resource.
val lastPositionInResource = positionsInResource.lastOrNull()?.locations?.position
```

**Reply by paulocoutinhox** [reply]
Perfect, thanks.

**Reply by paulocoutinhox** [reply]
Only more one thing @mickael-menu.

What is the best methods to:

- Go to a chapter
- Go to a page

**Reply by mickael-menu** [reply]
1. To go to a chapter, use the `navigator.go(link)` API using any link that you have in `publication.tableOfContents`, or `publication.readingOrder`.

3. Given a target `position` number, you can do:

```kotlin
val positions = publication.positions()
if (!positions.isEmpty()) {
    val index = (position - 1).coerceIn(0, positions.size - 1)
    navigator.go(positions[index])
}
```

**Reply by paulocoutinhox** [reply]
Perfect, thanks.

**Reply by rorodeveloper** [reply]
@mickael-menu 
> publication.positions.count, but they are not "screen page numbers"

Is there any way to get the actual "screen page numbers"?

**Reply by mickael-menu** [reply]
No, because it would require pre-rendering all the resources of the book and it changes every time the content is reflowed, for example when changing the font size or rotating the screen.
